### ISSUE TITLE
optimize string functions for STRANDS

### ISSUE THREAD

**Issue by nigoroll** [varnishcache/varnish-cache#3218]
Ref 0c3ec169282427e629f211dc17c4afd1df2bb9da

Here we went from
```
*b++ = (char)toupper(*p++);
```
to
```
VSB_putc(vsb, (char)toupper(*p));
```

now that we use VSBs more (which I appreciate), should be inline `VSB_putc()` and `VSB_put_byte()` to reduce overhead?

**Comment by bsdphk** [comment]
Only if we can measure a difference in some microbenchmark ?

**Comment by nigoroll** [comment]
So, I've run some benchmarks...

## TL;DR

* the WS_VSB does have an impact
* with inlining we can almost compensate for it

## summary

* the code before the switch to the new `WS_VSB` (see 0c3ec169282427e629f211dc17c4afd1df2bb9da) was fastest in all cases
* for gcc with `-O2` (which I assume a typical production build will use)
  * VSB_putc/WS_VSB is ~3x slower than the original code
  * the same with inlining is only ~10% slower than the original code (= ~2.8x faster than no inlining)
* for a developer build (without compiler optimizations and inlining)
  * VSB_putc/WS_VSB is ~4.5x slower than the original code
  * and inlining does not help because it is disabled
* for `-O6` which I tried out of curiosity
  * VSB_putc/WS_VSB also is ~3x slower than the original code (with both being faster than `-O2`)
  * inlining only gives a ~40% improvement and results in slower code than with `-O2`

## methodology

* to benchmark std.tolower(), I have used `vtc.benchmark()` from #3163 / https://github.com/varnishcache/varnish-cache/pull/3163/commits/902235a7be4244cbf0dac7524fa5a51b6eab7848 which uses the proposed `VRT_call`
* the vtc https://github.com/nigoroll/varnish-cache/commit/20da8cb6eb60707043c121423fa42660da2e080e takes a header of 3780 bytes length and repeatedly calls `std.tolower` on it. (the length is completely arbitrary except that I did not want to use too short a string for relevance)
* `if (std.tolower(..) == "")` is called 10 times in a vcl sub. The comparison is used to avoid the result to be discarded by any compiler optimization.
* This sub with the 10 `std.tolower` is timed using `vtc.benchmark()` with 10000 iterations which has been found to produce quite stable results
* three compiler settings and three different variants of the code have been tried
* the vtc has been run multiple times manually to check if results compare broadly (which they do). Typical results have been picked for the result tables.

## on the inlined code

Please consider the inlining test a hack. *It's ugly* due to massive `vsb.h` pollution. If we decided to go for inlining, we would need to invest some time to make the inlinable code cleaner.

## code used

in the tables below, I will refer by *what* to the following commits, which basically are in the order of the table:
* the code before WS_VSB / 0c3ec169282427e629f211dc17c4afd1df2bb9da.
* the code in master as of now
* experimental and crude vsb inlining

what | commit
-------| --------------------
*b++ = (char)tolower(*p++); | https://github.com/nigoroll/varnish-cache/commit/892598fe4e3e73addf79df591ea294af40d758e9
VSB_putc(vsb, (char)tolower(*p)); | https://github.com/nigoroll/varnish-cache/commit/20da8cb6eb60707043c121423fa42660da2e080e
vsb inlined | https://github.com/nigoroll/varnish-cache/commit/1b0b420989ccdfdf7f6a704548c093b71b675446

## results

### "autogen.des": gcc -O0 --no-inline

what | 10x std.tolower
-------| --------------------
*b++ = (char)tolower(*p++); | 186.003us
VSB_putc(vsb, (char)tolower(*p)); | 839.131us
vsb inlined | 852.543us

### CC=gcc CFLAGS='-O2' ./configure

what | 10x std.tolower
-------| --------------------
*b++ = (char)tolower(*p++); | 93.308us
VSB_putc(vsb, (char)tolower(*p)); | 292.152us
vsb inlined | 103.339us

### CC=gcc CFLAGS='-O6' ./configure

what | 10x std.tolower
-------| --------------------
*b++ = (char)tolower(*p++); | 63.723us
VSB_putc(vsb, (char)tolower(*p)); | 198.259us
vsb inlined | 140.219us

**Comment by nigoroll** [comment]
The other cases replaced with a WS_VSB look harmless with regard to this aspect (because formatting happens within VSB), so we might want to just revert 0c3ec169282427e629f211dc17c4afd1df2bb9da for the next release?

**Comment by Dridi** [comment]
Do we want a pair of `VSB_QUOTE_{UPP,LOW}ER` ?

**Comment by bsdphk** [comment]
So if I understand your benchmark, the worst case hit og not inlining is that a tolower call on a 3780 byte long string takes:

     (853{µs}-63{µs}) / 100000 = 8{ns}

longer ?

**Comment by nigoroll** [comment]
As given in the table header, the numbers are for _10x std.tolower_ (including some overhead).
So the worst case hit per `std.tolower` call is:
* for `"autogen.des": gcc -O0 --no-inline`: (839µs - 186µs) / 10 = 65µs
* for `CC=gcc CFLAGS='-O2' ./configure`: (292µs - 93µs) / 10 = 20µs
* for `CC=gcc CFLAGS='-O6' ./configure`: (198µs - 63 µs) / 10 = 13µs

**Comment by nigoroll** [comment]
> Do we want a pair of `VSB_QUOTE_{UPP,LOW}ER` ?

obviously that should remove the performance hit as well.

This is a classic case of keeping the inner loop efficient.

**Comment by Dridi** [comment]
Maybe a benchmark for the quote approach could help?

**Comment by bsdphk** [comment]
Re VSB_QUOTE_{UPPER|LOWER}:

Right now VSB doesn't come anywhere near the entire codepage/unicode-debacle.  If we were to implement them we would have to make it explicitly about ASCII-only, which feels inappropriate, given VSB's pedigree and design.

Thanks for clarifying the benchmark timings, I had a hard time making sense of it.

I agree that those times are a bit on the high side, but I do not find them excessive for a 3.5kB header.

I am defensive about VSB retaining its API isolation and VSB_put_byte(), (misnamed, should be lower-case) is firmly part of the implementation and not the API and only inlining VSB_putc() wouldn't do anything to speed things up.

**Comment by nigoroll** [comment]
So VSB is low overhead for anything _but_ external `VSB_putc()` calls like `VSB_printf()` or `VSB_cat()` for which the compiler can inline and optimize out a lot of code for the `VSB_putc()`s.
Also I do not want to over-stress the relevance of this performance degression, but I still wonder if
* we should just undo the one case with the performance penalty
* or if we could have an interface where some buffer region is delegated to the caller, similar to WS_Reserve()

**Comment by nigoroll** [comment]
FTR, unrelated to this topic I did a less on a mailinglist archive today and, after a long time, read again [the very first varnish-dev posting](https://varnish-cache.org/lists/pipermail/varnish-dev/2006-February/005040.html). I understand that we have gone a long way since _we want to keep it (varnish) small and efficient at all cost_, but I like to view it the other way around:
There should be a very good reason to sacrifice performance. In my mind, that reason does not exist for the toupper/tolower code. The WS_VSB interface is nice, but we should not use it dogmatically, and the previous code was just fine.

**Comment by nigoroll** [comment]
I think we should just revert 0c3ec169282427e629f211dc17c4afd1df2bb9da for the next release.
`WS_VSB` is a good interface, but no size fits all and we do not gain much for toupper/tolower.

**Comment by nigoroll** [comment]
bugwash:
* revert 0c3ec169282427e629f211dc17c4afd1df2bb9da
* the functions should check each strand if it needs modification and keep it unmodified if not
* unclear if anything else should be done

**Comment by nigoroll** [comment]
Open question on _the functions should check each strand if it needs modification and keep it unmodified if not_:

this lead me to pondering how we should handle the mixed strands / strings cases: Avoiding copies in conversion function is all good and well, but right now we still collect strands into strings by copying, so that would appear the more relevant optimization.

Consider a case of a vmod function taking a STRANDS argument where vcc sees a mixture of strings and strands. In that case, we currently collect strands into a single string and pass all strings as a STRANDS argument. To illustrate, here is a variant of the return_strands call from v58.vtc:

```
        set resp.http.Hdr-8 =
            debug.return_strands(
		"{" +
                debug.return_strands(
                    req.url + "<-->" + req.url
                )
            ) + "}";
```

compiles to:

```
      VRT_SetHdr(ctx, &VGC_HDR_RESP_Hdr_2d_8,
        VRT_CollectStrands(ctx,
          Vmod_xyzzy_debug_Func.return_strands(ctx,
            &(struct strands){.n = 2, .p = (const char *[2]){
            "{",
            VRT_CollectStrands(ctx,
              Vmod_xyzzy_debug_Func.return_strands(ctx,
                &(struct strands){.n = 3, .p = (const char *[3]){
                VRT_r_req_url(ctx),
                "<-->",
                VRT_r_req_url(ctx)
                }}
              )
            )
            }}
          )
        ),
```

In order to make use of strands to the extent of actually avoiding copies, it appears to to me that we should first strands-wrap all strings and then pass a flattened version as a single STRANDS argument.

I prototyped

```
VCL_STRANDS
VRT_FlattenStrands(struct ws *ws, VCL_STRANDS s, ...)
```

such that the outer return_strands() from above would turn into:

```
#define ST(x) &(struct strands){.n=1, .p = (const char*[1]){p}}

Vmod_xyzzy_debug_Func.return_strands(ctx,
  VRT_FlattenStrands(ctx->ws,
    ST("{"),
    Vmod_xyzzy_debug_Func.return_strands(ctx, ...),
    ST("}"),
    vrt_magic_thring_end));
```

Do we want to follow this route?

**Comment by bsdphk** [comment]
Closing this issue as done.

VRT_FlattenStrands() is for all practical purposes the same as VRT_STRANDS_string()
